python/app/input_widget.py

```diff
@@ -187,7 +187,17 @@ def execute(self):
         else:
             with nested(self._stdout_redirect, self._stdin_redirect):
                 try:
-                    exec(python_code, globals(), self._locals)
+                    # in python3 a collections.ChainMap might be a cleaner solution
+                    # by using one dictionary for both locals and globals 
+                    # it gets around the weird way exec adds things to the global scope
+                    exec_scope = dict(globals(), **self._locals)
+                    exec(python_code, exec_scope, exec_scope)
+                    # need to check what has changed in exec_scope and update our locals
+                    existing_scope = dict(globals(), **self._locals)
+                    # this wont handle del, but that probably isn't used a lot interactivly
+                    for key in exec_scope:
+                        if (key in existing_scope and existing_scope[key] != exec_scope[key]) or key not in existing_scope:
+                            self._locals[key] = exec_scope[key]
```

Comment:
Could this be simplified with the `self._locals.update(exec_scope)`

Reply:
Maybe but that would throw everything from `globals()` into `self._locals`
If we do that it might be easier to initialize `self._locals` with `globals()` and run `exec(python_code, self._locals, self._locals)`

Reply:
I wonder if we even need the `globals()` added?
The Shotgun predefined vars probably should be enough. If I understand this correctly all we are doing by adding `globals()` is including any global variables defined within the `input_widget.py` module. I wonder if it would be preferable to keep the scope cleaner and only include the Shotgun predefined variables? What do you think?

Reply:
That's probably a good idea. Nothing from globals looks necessary.

Interactive python shell has these other dunders set, so maybe these should be set
```
__name__ = '__main__'
__doc__ = None
__package__ = None
```
![globals_test](https://user-images.githubusercontent.com/6559190/73568326-0bbe9080-441d-11ea-84ff-a0f98b739faf.png)

Reply:
Ah, good point, I think Python 3 has a few other bits as well:
<img width="1487" alt="python3_7 — _Users_philips1_source_code_tk-multi-pythonconsole_and_Fix_execution_scope_by_jhultgre_·_Pull_Request__4_·_shotgunsoftware_tk-multi-pythonconsole" src="https://user-images.githubusercontent.com/3777228/73640230-ce990f00-4665-11ea-83a9-2463bac93119.png">
I haven't had time unfortunately to look up the importance of these and see if they are really required. 
I think by inheriting them from the `globals()` though would give you some incorrect/odd things like `__file__` pointing to the input widget, so I think not inheriting the globals it's the right thing to do. I'd like to get this over for QA soon hopefully but I can't make any promises. Thanks for your hard work!

Reply:
I decided its probably good enough to leave as is. If we see any issues arise we can address them then.